src/Databases/Iceberg/DatabaseIceberg.cpp

```diff
@@ -0,0 +1,386 @@
+{
+namespace DatabaseIcebergSetting
+{
+    extern const DatabaseIcebergSettingsDatabaseIcebergCatalogType catalog_type;
+    extern const DatabaseIcebergSettingsString warehouse;
+    extern const DatabaseIcebergSettingsString catalog_credential;
+    extern const DatabaseIcebergSettingsString auth_header;
+    extern const DatabaseIcebergSettingsString auth_scope;
+    extern const DatabaseIcebergSettingsString storage_endpoint;
+    extern const DatabaseIcebergSettingsString oauth_server_uri;
+    extern const DatabaseIcebergSettingsBool vended_credentials;
+}
+namespace Setting
+{
+    extern const SettingsBool allow_experimental_database_iceberg;
+}
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int SUPPORT_IS_DISABLED;
+}
+
+namespace
+{
+    /// Parse a string, containing at least one dot, into a two substrings:
+    /// A.B.C.D.E -> A.B.C.D and E, where
+    /// `A.B.C.D` is a table "namespace".
+    /// `E` is a table name.
+    std::pair<std::string, std::string> parseTableName(const std::string & name)
+    {
+        auto pos = name.rfind('.');
+        if (pos == std::string::npos)
+            throw DB::Exception(ErrorCodes::BAD_ARGUMENTS, "Table cannot have empty namespace: {}", name);
+
+        auto table_name = name.substr(pos + 1);
+        auto namespace_name = name.substr(0, name.size() - table_name.size() - 1);
+        return {namespace_name, table_name};
+    }
+}
+
+DatabaseIceberg::DatabaseIceberg(
+    const std::string & database_name_,
+    const std::string & url_,
+    const DatabaseIcebergSettings & settings_,
+    ASTPtr database_engine_definition_)
+    : IDatabase(database_name_)
+    , url(url_)
+    , settings(settings_)
+    , database_engine_definition(database_engine_definition_)
+    , log(getLogger("DatabaseIceberg(" + database_name_ + ")"))
+{
+    validateSettings();
+}
+
+void DatabaseIceberg::validateSettings()
+{
+    if (settings[DatabaseIcebergSetting::warehouse].value.empty())
+    {
+        throw Exception(
+            ErrorCodes::BAD_ARGUMENTS, "`warehouse` setting cannot be empty. "
+            "Please specify 'SETTINGS warehouse=<warehouse_name>' in the CREATE DATABASE query");
+    }
+}
+
+std::shared_ptr<Iceberg::ICatalog> DatabaseIceberg::getCatalog(ContextPtr) const
+{
+    if (catalog_impl)
+        return catalog_impl;
+
+    switch (settings[DatabaseIcebergSetting::catalog_type].value)
+    {
+        case DB::DatabaseIcebergCatalogType::REST:
+        {
+            catalog_impl = std::make_shared<Iceberg::RestCatalog>(
+                settings[DatabaseIcebergSetting::warehouse].value,
+                url,
+                settings[DatabaseIcebergSetting::catalog_credential].value,
+                settings[DatabaseIcebergSetting::auth_scope].value,
+                settings[DatabaseIcebergSetting::auth_header],
+                settings[DatabaseIcebergSetting::oauth_server_uri].value,
+                Context::getGlobalContextInstance());
```

Comment:
Removed redundant context argument. Global context is used because we save this object into class and reuse afterwards, therefore we cannot use local context, as it will lead to "context expired".